components/fxa-client/src/oauth.rs

```diff
@@ -167,25 +166,51 @@ impl FirefoxAccount {
     /// Fetch an OAuth code for a particular client using a session token from the account state.
     /// This method doesn't support OAuth public clients at this time.
     ///
-    /// * `client_id` - OAuth client id.
-    /// * `scopes` - Space-separated list of requested scopes.
-    /// * `state` - OAuth state.
-    /// * `access_type` - Type of OAuth access, can be "offline" and "online.
+    /// * `auth_params` Authentication parameters
     pub fn authorize_code_using_session_token(
         &self,
-        client_id: &str,
-        scope: &str,
-        state: &str,
-        access_type: &str,
+        auth_params: AuthorizationParameters,
     ) -> Result<String> {
         let session_token = self.get_session_token()?;
+        let auth_request_params = if let Some(key_params) = auth_params.key_params {
+            let keys_jwk = base64::decode_config(key_params.keys_jwk, base64::URL_SAFE_NO_PAD)?;
+            let keys_jwk = String::from_utf8(keys_jwk)?;
+            let requested_scopes: HashSet<&str> = auth_params.scope.split_whitespace().collect();
+            let data: HashMap<String, ScopedKey> = self
+                .state
+                .scoped_keys
+                .iter()
+                .filter(|(scope, _)| requested_scopes.contains(&(*scope).as_ref()))
+                .map(|(key, value)| (key.clone(), value.clone()))
+                .collect();
+            let data = serde_json::to_string(&data)?;
+            let client_keys_flow = ScopedKeysFlow::with_random_key()?;
+            let keys_jwe = client_keys_flow.encrypt_keys_jwe(&keys_jwk, data.as_bytes())?;
+            AuthorizationRequestParameters {
+                client_id: auth_params.client_id,
+                scope: auth_params.scope,
+                state: auth_params.state,
+                access_type: auth_params.access_type,
+                code_challenge: Some(key_params.code_challenge),
+                code_challenge_method: Some(key_params.code_challenge_method),
+                keys_jwe: Some(keys_jwe),
+            }
+        } else {
+            AuthorizationRequestParameters {
+                client_id: auth_params.client_id,
+                scope: auth_params.scope,
+                state: auth_params.state,
+                access_type: auth_params.access_type,
+                code_challenge: None,
+                code_challenge_method: None,
+                keys_jwe: None,
+            }
```

Comment:
Nit: you could probably factor the `AuthorizationRequestParameters` construction outside of the `if-let` block, whereas the `if-let` block would return a `Option<keys_jwe type>`, but disregard this comment if the memory checker gives you a hard time doing this.